src/phoenix/server/api/types/Project.py

```diff
@@ -189,17 +188,63 @@ async def spans(
         if filter_condition:
             span_filter = SpanFilter(condition=filter_condition)
             stmt = span_filter(stmt)
+        sortable_field: Optional[SortableField] = None
+        sort_result: Optional[SpanSortResult] = None
         if sort:
-            stmt = sort.update_orm_expr(stmt)
-        else:
-            stmt = stmt.order_by(desc(models.Span.id))
-        stmt = stmt.limit(
-            SPANS_LIMIT
-        )  # todo: remove this after adding pagination https://github.com/Arize-ai/phoenix/issues/3003
+            sort_result = sort.update_orm_expr(stmt)
+            stmt = sort_result.stmt
+        if after:
+            node_identifier = NodeIdentifier.from_cursor(after)
+            if node_identifier.sortable_field is not None:
+                sortable_field = node_identifier.sortable_field
+                assert sort is not None  # todo: refactor this into a validation check
+                compare = operator.lt if sort.dir is SortDir.desc else operator.gt
+                if sort_result:
+                    stmt = stmt.where(
+                        compare(
+                            tuple_(sort_result.orm_expression, models.Span.id),
+                            (sortable_field.value, node_identifier.rowid),
+                        )
+                    )
+            else:
+                stmt = stmt.where(models.Span.id < node_identifier.rowid)
+        if first:
+            stmt = stmt.limit(
+                first + 1  # overfetch by one to determine whether there's a next page
+            )
+        stmt = stmt.order_by(desc(models.Span.id))
+        data = []
         async with info.context.db() as session:
-            spans = await session.stream_scalars(stmt)
-            data = [to_gql_span(span) async for span in spans]
-        return connection_from_list(data=data, args=args)
+            rows = await session.execute(stmt)
+            async for row in islice(rows, first):
+                span = row[0]
```

Comment:
There's unfortunately not a straightforward way to do this in Python to my knowledge since we don't know the number of items in `span_record` ahead of time.